compiler/rustc_resolve/src/lib.rs

```diff
@@ -3350,6 +3358,42 @@ impl<'a> Resolver<'a> {
         }
         None
     }
+
+    fn resolve_main(&mut self) {
+        let module = self.get_module(LocalDefId { local_def_index: CRATE_DEF_INDEX }.to_def_id());
+        let ident = Ident::with_dummy_span(sym::main);
+        let parent_scope = &ParentScope::module(module, self);
+
+        let name_binding = match self.resolve_ident_in_module(
+            ModuleOrUniformRoot::Module(module),
+            ident,
+            ValueNS,
+            parent_scope,
+            false,
+            DUMMY_SP,
+        ) {
+            Ok(name_binding) => name_binding,
+            _ => return,
+        };
+
+        let res = name_binding.res();
+        let is_import = name_binding.is_import();
+        let span = name_binding.span;
+        if let Res::Def(DefKind::Fn, _) = res {
+            if is_import && !self.session.features_untracked().imported_main {
+                feature_err(
+                    &self.session.parse_sess,
+                    sym::imported_main,
+                    span,
+                    "using an imported function as entry point `main` is experimental",
+                )
+                .emit();
```

Comment:
This error should be reported in `compiler/rustc_passes/src/entry.rs` only if the resolution result really ends up being used.
In other words
```rust
#[start] fn my_start(...) { ... }

use something as main;
```
is not an error because `my_start` is used as an entry point.

Reply:
Sure, i'll add this as a test too.